rmw_zenoh_cpp/src/rmw_service.cpp

```diff
@@ -0,0 +1,419 @@
+#include "rcutils/strdup.h"
+
+#include <rmw/validate_full_topic_name.h>
+#include "rmw/impl/cpp/macros.hpp"
+#include "rmw/error_handling.h"
+#include "rmw/event.h"
+#include "rmw/rmw.h"
+
+#include "rmw_zenoh_cpp/rmw_context_impl.hpp"
+#include "rmw_zenoh_cpp/identifier.hpp"
+
+#include "impl/type_support_common.hpp"
+#include "impl/service_impl.hpp"
+#include "impl/client_impl.hpp"
+
+extern "C"
+{
+
+/// CREATE SERVICE SERVER ======================================================
+// Create and return an rmw service server
+rmw_service_t *
+rmw_create_service(
+  const rmw_node_t * node,
+  const rosidl_service_type_support_t * type_supports,
+  const char * service_name,
+  const rmw_qos_profile_t * qos_profile)
+{
+  RCUTILS_LOG_INFO_NAMED("rmw_zenoh_cpp", "[rmw_create_service] %s", service_name);
+
+  // ASSERTIONS ================================================================
+  RMW_CHECK_ARGUMENT_FOR_NULL(node, nullptr);
+
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    node,
+    node->implementation_identifier,
+    eclipse_zenoh_identifier,
+    return nullptr
+  );
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(service_name, nullptr);
+  if (strlen(service_name) == 0) {
+    RMW_SET_ERROR_MSG("service name is empty string");
+    return nullptr;
+  }
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(qos_profile, nullptr);
+  RMW_CHECK_ARGUMENT_FOR_NULL(type_supports, nullptr);
+
+  // OBTAIN ALLOCATOR ==========================================================
+  rcutils_allocator_t * allocator = &node->context->options.allocator;
+
+  // VALIDATE SERVICE NAME =====================================================
+  int * validation_result = static_cast<int *>(allocator->allocate(sizeof(int), allocator->state));
+
+  rmw_validate_full_topic_name(service_name, validation_result, nullptr);
+
+  if (*validation_result == RMW_TOPIC_VALID
+      || qos_profile->avoid_ros_namespace_conventions) {
+    allocator->deallocate(validation_result, allocator->state);
+  } else {
+    RMW_SET_ERROR_MSG("service name is malformed!");
+    allocator->deallocate(validation_result, allocator->state);
+    return nullptr;
+  }
+
+  // OBTAIN TYPESUPPORT ========================================================
+  const rosidl_service_type_support_t * type_support = get_service_typesupport_handle(
+    type_supports, RMW_ZENOH_CPP_TYPESUPPORT_C
+  );
+
+  if (!type_support) {
+    type_support = get_service_typesupport_handle(type_supports, RMW_ZENOH_CPP_TYPESUPPORT_CPP);
+    if (!type_support) {
+      RCUTILS_LOG_INFO("%s", service_name);
+      RMW_SET_ERROR_MSG("type support not from this implementation");
+      return nullptr;
+    }
+  }
+
+  // CREATE SERVICE ============================================================
+  rmw_service_t * service = static_cast<rmw_service_t *>(
+    allocator->allocate(sizeof(rmw_service_t), allocator->state)
+  );
+  if (!service) {
+    RMW_SET_ERROR_MSG("failed to allocate rmw_service_t");
+    allocator->deallocate(service, allocator->state);
```

Comment:
Don't deallocate something that was not allocated.

Reply:
Updated: 23497d4430f890dc62fb9995c66fb2be8602b51e

(Sorry for the massive commit, it's kind of intertwined...)